File: leet_lint_code/lint_code/majority_element.cpp
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of lint code problem 46.

#ifndef LEETLINTCODE_LINTCODE_MAJORITYELEMENT_CPP
#define LEETLINTCODE_LINTCODE_MAJORITYELEMENT_CPP


class MajorityElement {
 public:
    int majorityNumber(vector<int> &nums) {
      int cnt = 1, majority = nums[0], i = 1, len = (int)nums.size();
      while (i < len) {
        if (majority == nums[i])
          ++ cnt;
        else {
          -- cnt;
          if (0 == cnt) {
            majority = nums[i];
            cnt = 1;
          }
        }

        ++ i;
      }
      return majority;
    }
};

#endif // LEETLINTCODE_LINTCODE_MAJORITYELEMENT_CPP

File: leet_lint_code/leet_code/java/MajorityElement.java
// Copyright © 2018 Hailin Gu. All rights reserved.
// License
// Author: leetcode
// This is a answer of leet code problem 169.
// Date: 2019.7.16

class MajorityElement {
    public int majorityElement(int[] nums) {
        int ans = nums[0];
        int cnt = 1;
        for (int i = 1; i < nums.length; ++ i) {
            if (nums[i] != ans) {
                cnt -- ;
            } else {
                cnt ++ ;
            }
            
            if (cnt == 0) {
                cnt = 1;
                ans = nums[i];
            }
        }
        
        return ans;
    }
}